### DISCUSSION TITLE
Rule to ensure callee of function returning Try<Foo, Throwable> handles the throwable

### DISCUSSION THREAD

**Discussion by efenderbosch** [detekt/detekt#3536] [Q&A] [answered]
Which of the `visit...` functions would be a good place to start to implement a `Rule` to fail on use of function returning a `Try<Foo, Throwable>` and the callee not handling the `Throwable`?

This is what I have so far (based off `IgnoredReturnValue`):

```kotlin
/* @requiresTypeResolution */
class DanglingTry(config: Config) : Rule(config) {

    override val issue = Issue(
        "DanglingTry",
        Severity.Defect,
        "callee of Try does not handle Throwable",
        Debt.TEN_MINS
    )

    override fun visitCallExpression(expression: KtCallExpression) {
        super.visitCallExpression(expression)
        if (bindingContext == BindingContext.EMPTY) return
        if (expression.isUsedAsExpression(bindingContext)) return
        val resolvedCall = expression.getResolvedCall(bindingContext) ?: return
        val returnType = resolvedCall.resultingDescriptor.returnType ?: return
        if (returnType.getJetTypeFqName(false) != "com.acme.Try") return

        val entity = Entity.from(expression)
        val signature = entity.signature
        // is the best way to inspect the callee to see if onFailure, mapErr, etc are called on the returned Try?
        if (signature.contains("onFailure") || signature.contains("mapErr")) return

        val messageText = expression.calleeExpression?.text ?: expression.text
        report(
            CodeSmell(
                issue,
                entity,
                message = "The call $messageText is returning a Try that does not handle the Throwable."
            )
        )
    }
}

```

**Comment by cortinico** [comment, marked as answer]
> Which of the visit... functions would be a good place to start to implement a Rule to fail on use of function returning a Try<Foo, Throwable> and the callee not handling the Throwable?

To answer your question, `visitCallExpression` is the correct function. You can take inspiration from the `ForbiddenMethodCall` rule:

https://github.com/detekt/detekt/blob/c5f65a9c8ed029f0382674371c070ec0f3071b28/detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenMethodCall.kt#L75-L81

>  to fail on use of function returning a Try<Foo, Throwable> and the callee not handling the Throwable?

Can you provide a snippet of what you're trying to achieve?

**Reply by efenderbosch** [reply]
I'll be coming back to this soon-ish, but we've tabled it temporarily.
